crypto3/libs/blueprint/include/nil/blueprint/components/systems/snark/plonk/verifier/dfri_verifier.hpp

```diff
@@ -0,0 +1,932 @@
+                    row += poseidon_instance.rows_amount;
+                }
+
+                for( std::size_t i = 0; i < fri_params.lambda; i++){
+                    poseidon_input = {poseidon_output.output_state[2], zero_var, zero_var};
+                    poseidon_output = generate_circuit(poseidon_instance, bp, assignment, poseidon_input, row);
+                    c.fri_xs.push_back(poseidon_output.output_state[2]);
+                    row += poseidon_instance.rows_amount;
+                }
+
+                std::vector<var> xs;
+                std::vector<var> xf;
+                for( std::size_t i = 0; i < fri_params.lambda; i++){
+                    typename constant_pow_component_type::input_type constant_pow_input = {c.fri_xs[i]};
+                    typename constant_pow_component_type::result_type constant_pow_output = generate_circuit(
+                        constant_pow_instance, bp, assignment, constant_pow_input, row
+                    );
+                    xs.push_back(constant_pow_output.y);
+                    row+= constant_pow_instance.rows_amount;
+
+                    constant_pow_input = {xs[i]};
+                    constant_pow_output = generate_circuit(
+                        constant_pow2_instance, bp, assignment, constant_pow_input, row
+                    );
+                    xf.push_back(constant_pow_output.y);
+                    row+= constant_pow2_instance.rows_amount;
+                }
+
+                std::vector<var> x_indices;
+                for( std::size_t i = 0; i < fri_params.lambda; i++){
+                    typename x_index_component_type::input_type x_index_input;
+                    x_index_input.x = xs[i];
+                    for( std::size_t j = 0; j < log2(fri_params.domain_size) - 1; j++ ){
+                        x_index_input.b.push_back(instance_input.merkle_tree_positions[i][j]);
+                    }
+                    typename x_index_component_type::result_type x_index_output  = generate_circuit(
+                        x_index_instance, bp, assignment, x_index_input, row
+                    );
+                    row += x_index_instance.rows_amount;
+                    x_indices.push_back(x_index_output.b0);
+                }
+
+                std::vector<var> negative_xs;
+                std::size_t selector = generate_gates(component, bp, assignment, instance_input);
+                for(std::size_t i=0; i< fri_params.lambda; i++){
+                    auto xs_copy = var(component.W(0), row, false);
+                    auto xs_neg  = var(component.W(1), row, false);
+                    negative_xs.push_back(xs_neg);
+                    bp.add_copy_constraint({xs_copy, xs[i]});
+                    assignment.enable_selector(selector, row);
+                    row++;
+                }
+
+                std::vector<std::pair<std::size_t, std::size_t>> ordered_eval_map;
```

Comment:
No need in code duplication. Move to constructor.

Reply:
Moved to dFRI component constructor